### ISSUE TITLE
pipenv install --system - unsure what it does

### ISSUE THREAD

**Issue by thernstig** [pypa/pipenv#3383]
https://pipenv.readthedocs.io/en/latest/basics/ reads:

> --system — Use the system pip command rather than the one from your virtualenv.

https://pipenv.readthedocs.io/en/latest/advanced/ reads:

> **Deploying System Dependencies**
You can tell Pipenv to install a Pipfile’s contents into its parent system with the --system flag:
> $ pipenv install --system
This is useful for managing the system Python, and deployment infrastructure (e.g. Heroku does this).

I am having a hard time understanding what it does with these two descriptions. `basic` says it uses the pip command from the system. Which I assume means whatever `which pip` points to when you do not have a venv activated?

`advanced` says in the headline "Deploying System Dependencies". This makes it sounds to me it copies the system's site-packages directories into the virtual environment. I assume that assumption is incorrect?

The rest of the text in `advanced` makes it sound exactly like I am assuming it does when reading the `basic` text. But it is worded differently between the two.

To summarize:
- Is my assumption correct that `pipenv install --system `installs into the same location as where `which pip` shows when you have not activated a venv?
- Could this text be made easier to understand in both `basic` and `advanced`?

Thank you for a nice tool.

**Comment by jxltom** [comment]
Your understanding is correct except for the assumption about ```Deploying System Dependencies``` which is actually deploying your pipfile/pipfile.lock's content to your system via system's pip. Maybe this title ```Deploying System Dependencies``` is not clear enough? It will be cool if you have time to make a PR for that.

**Comment by thernstig** [comment]
@jxltom thank you for confirming. I am not sure what would suite better unfortunately.

**Comment by jxltom** [comment]
No, it will make sure pipfile.lock is not out-of-date and only install the packages to your system without creating virtualenv.

**Comment by thernstig** [comment]
@jxltom Does it have to create a temporary virtual env to make sure that the Pipfile.lock is out of date? "under the hood", so to say?

I suppose the term "out-of-date" is not obvious to someone new to this.

**Comment by jxltom** [comment]
So the workflow is like using pipenv to lock your packages in your local develop environment, so you have a Pipfile and Pipfile.lock which is up-to-date. Then you could deploy it by ```pipenv install --deploy --system``` in your production server. The ```--deploy``` will make sure your packages are properly locked in Pipfile.lock since it will check the hashes.

**Comment by thernstig** [comment]
So what --deploy does is to check the contents of Pipfile and compare to Pipfile.lock, and if one has added packages to the Pipfile (and nothing else such as pipenv lock), then Pipfile.lock would be considered out-of-date?

When it is checking the hashes, does it have to have to in any way install into a virtual environment? I am unsure how it would compare hashes here and against what. Since Pipfile does not contains hashes, but Pipfile.lock does it has to do some kind of "magic" here to check the packages of Pipfile (unhashed) towards Pipfile.lock (with hashes).

Would adding a text explaining this could be helpful if I gave it a try?

**Comment by jxltom** [comment]
If you add/remove packages or edit packages version in Pipfile without locking again, the Pipfile.lock is out-of-date.

The hash in Pipfile.lock is the SHA256 of pipfile data. The pipenv will compare the hash in pipfile.lock and calculated current Pipfile hash to check wether the Pipfile.lock is up-to-date or not.

**Comment by thernstig** [comment]
@jxltom Thank you for the patience on this btw :)

The whole confusion from this stems from that I do a `/prefix/bin/pipenv install --deploy --system`. Where `prefix` comes from doing a `./configure --prefix=/some/path/python --with-ensurepip=upgrade && make && make install`

But `/prefix/bin/pipenv install --deploy --system` does **not** install into /prefix/lib/python2.7/site-packages/ like expected. It installs into a virtual environment. This is with using `pip-18.1` and  `pipenv-2018.11.26`.

So I am utterly confused here.

**Comment by jxltom** [comment]
@thernstig Open a fresh terminal and what's the output of ```which pip```?

**Comment by thernstig** [comment]
@jxltom I think your suggestion helped. `which python` shows `/usr/bin/python` and `which pip` shows `which: no pip in (/usr/local/bin:/usr/bin)`. I suppose the solution is to add the location of `/prefix/bin/` to PATH?

I just assumed that using `/prefix/bin/pipenv install --deploy --system` would use pip from the same prefix.

**Comment by thernstig** [comment]
I am unsure how to solve my issue. This is being built in a very minimal build system in a "bare" environment. I tried with prefixing my PATH with `prefix/bin` but that is a no-go. Reason being that PATH is empty before this. This makes all other tools used fail since the system tries to find them in PATH.

I then tried `pipenv install --deploy --system --python /prefix/python/bin/python` but that also did not work. it does not install them into `/prefix/lib/python2.7/site-packages/` which is what I want.

**Comment by jxltom** [comment]
If your system doesn't know which pip to use, I think pipenv is not smart enough to find it out. It's better to append your pip path to your PATH.

I don't think pipenv is creating virtualenv from your output. You mean ```'-r', '/tmp/pipenv-XVpJ_x-requirements/pipenv-88PXSt-requirement.txt'```? This is not virtualenv, this is just temporary file which is created by pipenv for installing packages. More specifically, this is a file converted from Pipfile and so that ```pip``` can understand.

**Comment by jxltom** [comment]
- ```pip install pipenv``` is just installing pipenv as a tool, you can even install pipenv by brew if you're in mac. So the pip here doesn't matter.
- ```pipenv install --system``` is to install packages to your system.
- ```pipenv install``` will create virtualenv and install packages to it.

Probably this explanation makes more sense?

Sure, if you think the docs are not clear enough, please go ahead and make a PR, we will be happy to review 😄

**Comment by thernstig** [comment]
@jxltom I will see if I can come up with something that is clear enough.

Your explanation is clear. Would it make sense for `/prefix/bin/pipenv install --deploy --system` to use pip from `/prefix/bin/pip` instead of from what is found in the path?

**Comment by jxltom** [comment]
The ```--deploy``` will install packages by your system's pip. Usually the system pip is the one which can be found by ```which pip```. It has nothing to do with where pipenv is installed.

**Comment by jxltom** [comment]
Sorry typo 😭 Glad to help!

**Comment by tapaswenipathak** [comment]
Hi folks: Writing a PR.

**Comment by thernstig** [comment]
@tapaswenipathak @frostming I am unsure how your merge request solved this issue. Did you tag the incorrect issue? This issue was about the `--system` option whereas the MR indicates you updated information about `--deploy`

**Comment by frostming** [comment]
@thernstig Oh, are you still having difficulty understanding the `--system` option?

**Comment by thernstig** [comment]
@frostming No, but others might :) Just was confused why it was closed with that particular MR, but I am fine generally closing this.
